kani-driver/src/call_cbmc.rs

```diff
@@ -87,18 +97,70 @@ impl KaniSession {
             // Done here because `--visualize` uses the XML format instead.
             cmd.arg("--json-ui");
 
-            // Spawn the CBMC process and process its output below
-            let cbmc_process =
-                self.run_piped(cmd).map_err(|_| anyhow::Error::msg("Failed to run cbmc"))?;
-            let output = process_cbmc_output(cbmc_process, |i| {
+            self.runtime.block_on(self.run_cbmc_piped(cmd, harness))?
+        };
+
+        Ok(verification_results)
+    }
+
+    async fn run_cbmc_piped(
+        &self,
+        mut cmd: TokioCommand,
+        harness: &HarnessMetadata,
+    ) -> Result<VerificationResult> {
+        if self.args.common_args.verbose() {
+            println!("[Kani] Running: `{}`", render_command(cmd.as_std()).to_string_lossy());
+        }
+        // Spawn the CBMC process and process its output below
+        let mut cbmc_process = cmd
+            .stdout(std::process::Stdio::piped())
+            .spawn()
+            .map_err(|_| anyhow::Error::msg("Failed to run cbmc"))?;
+
+        let start_time = Instant::now();
+
+        let res = if let Some(timeout) = self.args.harness_timeout {
+            tokio::time::timeout(
+                timeout.into(),
+                process_cbmc_output(&mut cbmc_process, |i| {
+                    kani_cbmc_output_filter(
+                        i,
+                        self.args.extra_pointer_checks,
+                        self.args.common_args.quiet,
+                        &self.args.output_format,
+                    )
+                }),
+            )
+            .await
+        } else {
+            Ok(process_cbmc_output(&mut cbmc_process, |i| {
                 kani_cbmc_output_filter(
                     i,
                     self.args.extra_pointer_checks,
                     self.args.common_args.quiet,
                     &self.args.output_format,
                 )
-            })?;
+            })
+            .await)
+        };
 
+        let verification_results = if res.is_err() {
+            // An error occurs if the timeout was reached
+
+            // Kill the process
+            cbmc_process.kill().await?;
```

Comment:
It might be difficult to test this in CI, but could you please locally run a test with `--cbmc-args --external-solver cadical` (or whichever solver you like), and perhaps also with `--cbmc-args --z3` on examples that time out and require killing? I am asking for this as CBMC does some signal handling and I have previously seen issues where we couldn't reliably kill CBMC or its child processes.

Reply:
I tested this scenario with cadical and z3.

For cadical, both the CBMC and cadical processes were terminated right after reaching the timeout.

For z3, the CBMC process was terminated, but the z3 one continued to run, apparently till completion.

I'm not sure if there's anything that can be done from the Kani side to deal with the orphaned z3 process.

Reply:
Would you mind filing an issue against CBMC with this observation? I agree this isn’t for Kani to fix, but perhaps we will finally manage to settle this problem on the CBMC side for this scenario looks like one that is finally one that is reproducible. Thank you!

Reply:
Never mind, I was able to reproduce this with CBMC alone. The crucial piece is to send SIGKILL to the CBMC process (which is what [`Child::kill`](https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.kill) does) and not SIGTERM.

I filed https://github.com/diffblue/cbmc/issues/8485